statsmodels/regression/recursive_ls.py

```diff
@@ -267,6 +266,176 @@ def __init__(self, model, params, filter_results, cov_type='opg',
+        ==============================================================================
+        >>> T_test.effect
+        -1829.2025687192481
+        >>> T_test.sd
+        455.39079425193762
+        >>> T_test.tvalue
+        -4.0167754636411717
+        >>> T_test.pvalue
+        0.0015163772380899498
+
+        Alternatively, you can specify the hypothesis tests using a string
+
+        >>> from statsmodels.formula.api import ols
+        >>> dta = sm.datasets.longley.load_pandas().data
+        >>> formula = 'TOTEMP ~ GNPDEFL + GNP + UNEMP + ARMED + POP + YEAR'
+        >>> results = ols(formula, dta).fit()
+        >>> hypotheses = 'GNPDEFL = GNP, UNEMP = 2, YEAR/1829 = 1'
+        >>> t_test = results.t_test(hypotheses)
+        >>> print(t_test)
+                                     Test for Constraints
+        ==============================================================================
+                         coef    std err          t      P>|t|      [0.025      0.975]
+        ------------------------------------------------------------------------------
+        c0            15.0977     84.937      0.178      0.863    -177.042     207.238
+        c1            -2.0202      0.488     -8.231      0.000      -3.125      -0.915
+        c2             1.0001      0.249      0.000      1.000       0.437       1.563
+        ==============================================================================
+
+        See Also
+        ---------
+        tvalues : individual t statistics
+        f_test : for F tests
+        patsy.DesignInfo.linear_constraint
+        """
+        from statsmodels.tools.tools import recipr
+        from statsmodels.stats.contrast import ContrastResults
+        from patsy import DesignInfo
+        names = self.model.data.param_names
+        LC = DesignInfo(names).linear_constraint(r_matrix)
+        r_matrix, q_matrix = LC.coefs, LC.constants
+        num_ttests = r_matrix.shape[0]
+        num_params = r_matrix.shape[1]
+
+        # Check we don't have conflicting `loc` and `cov_p`
+        if loc is not None and cov_p is not None:
+            raise ValueError('Cannot use `loc` in combination with `cov_p`.')
+
+        if num_params != self.params.shape[0]:
+            raise ValueError('r_matrix and params are not aligned')
+        if q_matrix is None:
+            q_matrix = np.zeros(num_ttests)
+        else:
+            q_matrix = np.asarray(q_matrix)
+            q_matrix = q_matrix.squeeze()
+        if q_matrix.size > 1:
+            if q_matrix.shape[0] != num_ttests:
+                raise ValueError("r_matrix and q_matrix must have the same "
+                                 "number of rows")
+
+        if use_t is None:
+            # switch to use_t false if undefined
+            use_t = (hasattr(self, 'use_t') and self.use_t)
+
+        _t = _sd = None
+
+        # If we've been given a location
```

Comment:
It looks like most of the code is copied from elswhere, with the next bits looking less familiar than the others.  Can any of it be shared?

Reply:
Yeah, this is basically copied from `model.py::LikelihoodModelResults::t_test`. But the `loc` parameter is added, and the `df_resid` computation below is different. Maybe there's a better way to handle it. Two options:

- Refactor the base `t_test` to allow passing more things in as parameters. The problem is that I can't see many other use cases that would require that.
- In this `t_test`, temporarily override `self.params` and `self.df_resid`, call the parent method, and then set them back to were they were. This seems like a bummer to me since it's not very "clean", but I suppose it might be better?

Other options?

Reply:
My suggestion would be to make repeated code into helper method/functions.  This isn't a hill I'd make a stand on, but copy/pasted code is big concern so there might be some upside to encouraging/forming good habits/norms.